### DISCUSSION TITLE
blog/the-uphill-battle-of-memoization

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#99] [Announcements]
# blog/the-uphill-battle-of-memoization

Exploring why React.memo shouldn't be your first choice for performance optimizations

https://tkdodo.eu/blog/the-uphill-battle-of-memoization

**Comment by lgenzelis** [comment]
Excellent read!! You’re totally right about the *why* being missing in previous takes on this matter.

**Comment by nhatncb** [comment]
Do you think whether it's good to use React.memo like this with Record<string, string>:
memo(Component, (prevProps, props) => JSON.stringify(prevProps) === JSON.stringify(props))

**Reply by TkDodo** [reply]
it's explicitly mentioned in the docs that this is not good:

https://react.dev/reference/react/memo#specifying-a-custom-comparison-function

> Avoid doing deep equality checks inside arePropsEqual unless you are 100% sure that the data structure you’re working with has a known limited depth. Deep equality checks can become incredibly slow and can freeze your app for many seconds if someone changes the data structure later.

**Reply by shmuel-taub** [reply]
I think you are not guaranteed that the string output will be always the same (the order of keys may change)

**Comment by rimzzlabs** [comment]
Awesome reads, thank you for sharing.

You mentioned about the _**stable <p/\>**_ tag there.
What about children function? like [headless ui](https://headlessui.dev) does,even if there's nothing change on the props, the function always invoked on every re-render?

**Reply by TkDodo** [reply]
The function is a prop passed to the component. If the component is memoized, the function needs to be referentially stable, so you'd need to wrap it in useCallback. But I doubt that a component lib has memoized components that take functions as props :)

**Comment by WojciechMatuszewski** [comment]
Thank you for writing this blog post. It is very informative.

In addition to what you wrote, I find that making it explicit that a given component is memoized helps a ton. How might one do that? In my case, I will rename the original component to something akin to 'MemoizedXX', where 'XX' is the original name of the component.

Of course, this is not bulletproof, but I found that this makes other developers pay more attention when refactoring code.

**Comment by ignlopezsanchez** [comment]
So in the example of composition with children, the reference of children is not always the same?
I mean, is the same case in the example you have explained about children and memo. If we cannot trust the children with memo, why does it work in composition?
Thanks!

**Reply by ignlopezsanchez** [reply]
I missunderstood. In both examples the behavior is the same. If App rerenders, of course the reference of children changes.
But with the memo example it is true I had not thought about this. The reference of children changes so it breaks react memo.
The same in the first example if App rerenders the children component rerenders.

**Reply by TkDodo** [reply]
> The reference of children changes so it breaks react memo

exactly 👍

**Comment by subhadipPal** [comment]
Agreed, but one thing I would like to ask, 
const MemoizedComponent = memo(SomeComponent, arePropsEqual?)
arePropsEqual can be used to control the memoization/ re-rendering right based on the props ?

**Reply by TkDodo** [reply]
Yes but you have to include all props in the comparison. How is this helpful unless you have no functions and deep compare all objects (which can in itself be slower than just rendering the component)

**Comment by dangkhoipro** [comment]
Thank you!

**Comment by lucasheriques** [comment]
incredible read, thank you. The company I work for has been facing challenges with React applications - we just learned not everyone has M1 macbooks. Our app was so slow that it even crashed on an older device.

Current, we just employ the strategy to [memo everything](https://attardi.org/why-we-memo-all-the-things/) and worry about it later (turns out we just reached that). 

We run a very data-heavy application, with many nested objects, big data tables, etc. Since it's a very large codebase as well, it makes it harder to do much changes at once. I'll however try to look up the most expensive sections and try other approaches discussed in this article, so I appreciate it a lot.

Do you have any tips on actually measuring React performance? I'm embarrassed to say I haven't had the opportunity to tackle these performance issues, so I still have many things to learn.

**Reply by TkDodo** [reply]
the react profiler in the react devtools is pretty good

**Comment by bacher** [comment]
I can suggest another solution – memoization but explicit. You can memorize JSX actually, and have explicit revalidation list:
```typescript
function App() {
  const [state, setState] = React.useState({
    table1Data: [],
    table2Data: [],
  })

  const table1 = useMemo(() => {
    return <Table1 data={state.table1Data} />;
  }, [state.table1Data]);

  const table2 = useMemo(() => {
    return <Table2 data={state.table2Data} />;
  }, [state.table2Data]);

  return (
    <div>
      {table1}
      {table2}
      <SummaryBar />
    </div>
  )
}
```
So only changed table will be rerendered.

It's quite reliable solution, because developers already use `useMemo` a lot, and it's more predictable.

**Reply by TkDodo** [reply]
I actually like that a lot, I think its way better than having `memo` on the component. The eslint rules can even tell you when you add a dependency that is changing on every render. Only downside is you'd have to do this for every call-side, but maybe a custom hook can do 😅

**Reply by ryan-zayne** [reply]
@TkDodo So how then would you approach this with a custom hook? I'm genuinely interested

**Reply by TkDodo** [reply]
I don't understand the question? The article isn't about custom hooks, but about the `memo` higher order component.

Can you show an example of what you mean?

**Reply by ryan-zayne** [reply]
>Only downside is you'd have to do this for every call-side, but maybe a custom hook can do 😅
>

@TkDodo oh no, I was actually referring this statement you made😅

**Reply by TkDodo** [reply]
ooh okay, something like:

```
const Table 1 = () => ...

export const useTable1 = (data) => {
  return useMemo(() => {
    return <Table1 data={data} />
  }, [data])
}
```

note that that Table itself isn't exported, only the hook is. But you'd still run the risk of this hook being called with unstable inputs:

```
useTable1({ ...data, someField: true })
```

so yeah, I don't think I like abstracting it to a custom hook that much.

**Reply by ryan-zayne** [reply]
Fair enough actually

**Comment by jonybekov** [comment]
React Forget doesn't seem to be ready in near future.
https://www.reddit.com/r/reactjs/comments/16nnh4z/comment/k1jbr4t/

**Comment by ershisan99** [comment]
Great read, thanks! I have a question though:
How would you approach building an application with component driven design in mind? At my job we rely heavily on storybook because our data is somewhat complex and getting to a specific state of the application might take a while. It's a hot mess right now and I've been thinking about optimizing it a bit lately. One of the issues I'm facing is the usage of redux in our apps. Right now in some places the state is pulled directly from the store in the component, so closest to the rendering and that is what you are suggesting. However, component driven approach suggests we pull the state up to the page level and pass it down as props, which is where extra rerenders could occur. 
It might be a good idea to just mock the store, but, as I've mentioned, our apps are poorly designed so it's actually really hard to mock the state needed. Also it would mean losing the automatic controls based on props in storybook.
 Is memo/useMemo the only way for this approach or am I missing something?

**Reply by TkDodo** [reply]
> However, component driven approach suggests we pull the state up to the page level and pass it down as props, which is where extra rerenders could occur.

I'm not a fan of this split in general. I don't think it has much to do with `memo` in general though, it's more about which dependencies your components need. FWIW, we use storybook for our components as well, and those components do their own data fetching with React Query and read data directly from external stores like zustand as well. It works fine. Instead of passing things as a prop, you'd just wrap the component in the correct provider tree to "provide" data it needs. It's just a different form of dependency injection.

I have a bit of content in another blogpost here: https://tkdodo.eu/blog/react-query-as-a-state-manager#a-note-on-separation-of-concerns

**Comment by ashuvssut** [comment]
I haven't used zustand in my project. I guess the state I get from redux useSelector does the same thing that zustand helps with.

-----------------

Regarding my question...

I loved the "short circuit" concept you mentioned when using "content up" method

In you first code snippet that demonstrates that "content up" method, what if the `App` component has a state. If that state changes, then the `ExpensiveTree` re-renders... Seems like we have to use `React.memo` here, else we would just be doing the uphill battle of moving `ExpensiveTree` up (Content up)

Please let me know if I have misunderstood something. I currently feel like, in real complex projects with deep component trees, content up could not be useful as there could be a lot a parent components (having their own state) above `ExpensiveTree` component. If any of those parent component re renders `ExpensiveTree` re-renders too

`React.memo` seems the only solution to skip re-render of the `ExpensiveTree` component

**Reply by TkDodo** [reply]
> I haven't used zustand in my project. I guess the state I get from redux useSelector does the same thing that zustand helps with.

yes, exactly. One notable difference is that you don't need to have one large global store, but can have multiple small zustand stores. We do one per feature for example.

> In you first code snippet that demonstrates that "content up" method, what if the App component has a state. If that state changes, then the ExpensiveTree re-renders... Seems like we have to use React.memo here, else we would just be doing the uphill battle of moving ExpensiveTree up (Content up)

Yes, it would, but if that is an often changing state, you can apply the same concept again:

```
function App({children }) {
  const [oftenChangingState, setOftenChangingState] = useState()
  return (
    <ColorPicker>
      {children}
    </ColorPicker>
  )
}
```

depending on where that state goes, you could even lift the whole `ColorPicker` into `children`. You can't do that if `App` is the top level component, but in that case, I'd just suggest to not have often changing state at the top of your react tree ;)

**Reply by artem-malko** [reply]
But in a real life we have big applications with a lot of components. And do not forget about specific business logic, which can be changed every day. What is easier: recompose components every time or just write memo and you're ready to go?)

>  I'd just suggest to not have often changing state at the top of your react tree ;)

An expensive tree can be somewhere deep in a tree, where could be `oftenChangingState `) 

Looks like we have to think a lot about it. But what if we'll just forget about it, what if we will use memo everywhere, without thinking?

**Comment by artem-malko** [comment]
THX for the article) 

children in memo props — this blown my mind) I have to recap everything about React))

**Comment by niltonxp2** [comment]
Thank you for sharing!! 

there's a typo here: "...children we're passing **the the** memoized component into useMemo..."

**Reply by TkDodo** [reply]
thanks. Please open a PR :)

**Comment by WixeI** [comment]
A bit late to the party. Great read! I just have one question:

In your example, the Zustand solution works really well. But what would your approach be if the component that renders Table1, Table2, and Summary wasn’t a page, but a reusable component instead?

I get that this exact scenario might not make much sense, but it’s easier to stick with it than to explain the one I’m working on (which involves a reusable audio visualizer).

The main issue is that Zustand isn’t really designed for multiple isolated instances of the same store (at least not without some extra work). Curious how you'd handle that.

**Reply by TkDodo** [reply]
I have another blog post for that:

- https://tkdodo.eu/blog/zustand-and-react-context

**Comment by shmuel-taub** [comment]
Seems like passing the arePropsEqual and checking only the relevant props is the right solution to the most of the problems

**Reply by TkDodo** [reply]
Leaving out props is a dangerous practice, because “why would they not matter”? A component needs to re-render if one its props changes, otherwise, why would it get that prop if it doesn’t change anything?

If the answer is “because they are functions”, you’re ignoring that functions can close over values and if you omit them you’ll get stale closure problems. I’ve written about that here: https://tkdodo.eu/blog/hooks-dependencies-and-stale-closures#another-example
